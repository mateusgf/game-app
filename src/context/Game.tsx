import React, { createContext, useState, useContext } from "react";

import getGames from "../api/requests/game/getGames";
import createGameAPI from "../api/requests/game/createGame";
import joinGameAPI from "../api/requests/game/joinGame";
import getGameByIdAPI from "../api/requests/game/getGameById";
import getRoundsByGameIdAPI from "../api/requests/round/getRoundsByGameId";
import createRoundAPI from "../api/requests/round/createRound";
import updateRoundAPI from "../api/requests/round/updateRound";
import { usePlayer } from "./Player";
import { GameType, GameRound, CreateRoundType, UpdateRoundType } from "../types";

interface IContextProps {
  isLoading: boolean;
  fetchGames: Function;
  createGame: Function;
  joinGame: Function;
  games: GameType[];
  currentGame: GameType;
  getGameById: Function;
  fetchRoundsByGameId: Function;
  gameRounds: GameRound[];
  createRound: Function;
  updateRound: Function;
}

const GameContext = createContext({} as IContextProps);

export default function GameProvider({ children }: any) {
  const { setCurrentPlayer } = usePlayer();
  const [games, setGames] = useState([]);
  const [gameRounds, setGameRounds] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const [currentGame, setCurrentGame] = useState<GameType>({
    id: 0,
    hostNickname: "",
    guestNickname: "",
    numberOfRounds: 0,
  });

  const fetchGames = async () => {
    const res = await getGames();
    setGames(res);
  };

  const getGameById = async (id: number) => {
    setIsLoading(true);
    try {
      const game = await getGameByIdAPI(id);
      setCurrentGame(game);
      return game;
    } catch (error: any) {
      //@TODO: threat error to display friendly message to user
      alert(error);
    }
    setIsLoading(false);
  };

  const createGame = async (hostNickname: string, numberOfRounds: number) => {
    setIsLoading(true);
    try {
      const game = await createGameAPI(hostNickname, numberOfRounds);
      setCurrentGame(game);
      return game;
    } catch (error: any) {
      //@TODO: threat error to display friendly message to user
      alert(error);
    }
    setIsLoading(false);
  };

  const joinGame = async (gameId: number, nickname: string) => {
    setIsLoading(true);
    try {
      const game = await joinGameAPI(gameId, nickname);
      setCurrentGame(game);
      setCurrentPlayer({ nickname });
      return game;
    } catch (error: any) {
      //@TODO: threat error to display friendly message to user
      alert(error);
    }
    setIsLoading(false);
  };

  const fetchRoundsByGameId = async (gameId: number) => {
    setIsLoading(true);
    try {
      const rounds = await getRoundsByGameIdAPI(gameId);
      setGameRounds(rounds);
      return rounds;
    } catch (error: any) {
      //@TODO: threat error to display friendly message to user
      alert(error);
    }
    setIsLoading(false);
  };

  const createRound = async ({ gameId, hostAction, guestAction }: CreateRoundType) => {
    console.log("createRound", { gameId, hostAction, guestAction });
    setIsLoading(true);
    try {
      const rounds = await createRoundAPI({ gameId, hostAction, guestAction });
      setGameRounds(rounds);
      return rounds;
    } catch (error: any) {
      //@TODO: threat error to display friendly message to user
      alert(error);
    }
    setIsLoading(false);
  };

  const updateRound = async ({
    id,
    gameId,
    hostAction,
    guestAction,
  }: UpdateRoundType) => {
    console.log("updateRound", { id, gameId, hostAction, guestAction });
    setIsLoading(true);
    try {
      const rounds = await updateRoundAPI({
        id,
        gameId,
        hostAction,
        guestAction,
      });
      setGameRounds(rounds);
      return rounds;
    } catch (error: any) {
      //@TODO: threat error to display friendly message to user
      alert(error);
    }
    setIsLoading(false);
  };

  return (
    <GameContext.Provider
      value={{
        isLoading,
        fetchGames,
        createGame,
        joinGame,
        games,
        currentGame,
        getGameById,
        gameRounds,
        fetchRoundsByGameId,
        createRound,
        updateRound,
      }}
    >
      {children}
    </GameContext.Provider>
  );
}

export function useGame() {
  const context = useContext(GameContext);
  const {
    isLoading,
    fetchGames,
    createGame,
    joinGame,
    games,
    currentGame,
    getGameById,
    gameRounds,
    fetchRoundsByGameId,
    createRound,
    updateRound,
  } = context;

  return {
    isLoading,
    fetchGames,
    createGame,
    joinGame,
    games,
    currentGame,
    getGameById,
    gameRounds,
    fetchRoundsByGameId,
    createRound,
    updateRound,
  };
}
