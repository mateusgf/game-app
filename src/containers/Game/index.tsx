import { useEffect } from "react";
import { useHistory } from "react-router-dom";
import { useGame } from "../../context/Game";
import { usePlayer } from "../../context/Player";
import { routeCodes } from "../../constants/routes";
import PlayerScreen from "../../components/Organism/PlayerScreen";

const Game = (props: any) => {
  const history = useHistory();
  const { currentPlayer } = usePlayer();
  const {
    currentGame,
    getGameById,
    fetchRoundsByGameId,
    gameRounds,
    createRound,
    updateRound,
  } = useGame();

  useEffect(() => {
    const load = async () => {
      const id = props.match.params.id;
      if (!currentPlayer.nickname) {
        history.push(routeCodes.HOME);
      } else if (currentPlayer.nickname && id) {
        await getGameById(id);
        await fetchRoundsByGameId(id);
      }
    };
    load();

    // @TODO: put in place websock to avoid multiple requests to api to retrieve the last rounds for this game
    const intervalRef = setInterval(() => {
      fetchRoundsByGameId(props.match.params.id);
      if (!canHaveMoreRounds()) {
        clearInterval(intervalRef);
        getBestOfRounds();
      }
    }, 2000);

    return () => {
      clearInterval(intervalRef);
    };
  }, []);

  const isHostPlayer = () => {
    return currentGame.hostNickname === currentPlayer.nickname;
  };

  const playAction = async (action: string, isHostPlayer: boolean) => {
    // console.log('action triggered: ', { action, isHostPlayer });
    const lastRound = gameRounds[0];
    if (gameRounds.length === 0 || (lastRound.hostAction && lastRound.guestAction)) {
      // First action
      if (isHostPlayer) {
        await createRound({ gameId: currentGame.id, hostAction: action });
      } else {
        await createRound({ gameId: currentGame.id, guestAction: action });
      }
    } else {
      // check if last round created is missing an action either from HOST or GUEST
      // console.log('here', lastRound);
      if (lastRound && !lastRound.hostAction) {
        // Missing host action
        await updateRound({
          id: lastRound.id,
          gameId: currentGame.id,
          hostAction: action,
        });
      } else if (lastRound && !lastRound.guestAction) {
        await updateRound({
          id: lastRound.id,
          gameId: currentGame.id,
          guestAction: action,
        });
      }
    }
    await getGameById(currentGame.id);
  };

  const isAvailableToPlayAction = (isHostPlayer: boolean) => {
    // if missing action from oponent
    const lastRound = gameRounds[0];

    if (lastRound && isHostPlayer && !lastRound.hostAction) {
      // Last round played by guest. Enable host to play
      return true;
    } else if (lastRound && !isHostPlayer && !lastRound.guestAction) {
      // Last round played by host. Enable guest to play
      return true;
    } else if (
      lastRound &&
      lastRound.guestAction &&
      lastRound.hostAction &&
      isHostPlayer
    ) {
      // TODO: Those two last if will enable only HOST to start the game. Once websocket is in place either HOST or GUEST can trigger the first action.
      // This is to avoid them calling the first action simultaneously
      return true; // enable only host to go first
    } else if (gameRounds && gameRounds.length === 0 && isHostPlayer) {
      return true; // enable only host to go first
    }
    // No rounds found. Either Host or Guest can play first
    return false;
  };

  const getResultMessageForLastRound = () => {
    const lastRound = gameRounds?.[0];
    if (lastRound && lastRound.winnerNickname && lastRound.winnerAction !== "draw") {
      return `Winner of last turn was: ${lastRound.winnerNickname}`;
    } else if (lastRound && lastRound.winnerAction === "draw") {
      return `Last turn was a draw`;
    }
  };

  const canHaveMoreRounds = () => {
    const lastRound = gameRounds?.[0];
    if ((gameRounds && gameRounds.length) < currentGame.numberOfRounds) {
      return true;
    } else if (
      (gameRounds && gameRounds.length) === currentGame.numberOfRounds &&
      stillMissingActionToEndRound()
    ) {
      return true;
    }

    return false;
  };

  const stillMissingActionToEndRound = () => {
    const lastRound = gameRounds?.[0];
    return !lastRound.guestAction || !lastRound.hostAction;
  };

  const getBestOfRounds = () => {
    if (currentGame && gameRounds) {
      const winsByNickname = {
        [currentGame.hostNickname]: 0,
        [currentGame.guestNickname]: 0,
      };

      gameRounds
        .filter((round) => round.winnerAction !== "draw")
        .forEach((round) => {
          const winner = round.winnerNickname;
          winsByNickname[winner as string] += 1;
        });

      return winsByNickname;
    }
  };

  return (
    <>
      {JSON.stringify(gameRounds)}
      <PlayerScreen
        currentGame={currentGame}
        currentPlayer={currentPlayer}
        isHostPlayer={isHostPlayer()}
        gameRounds={gameRounds}
        roundsWinner={getBestOfRounds()}
        canHaveMoreRounds={canHaveMoreRounds()}
        resultMessageForLastRound={getResultMessageForLastRound()}
        isAvailableToPlayAction={isAvailableToPlayAction}
        playAction={playAction}
      />
    </>
  );
};

export default Game;
