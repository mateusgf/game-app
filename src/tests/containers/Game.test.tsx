import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import Game from "../../containers/Game";

const mockCurrentGame = {
  id: 14,
  hostNickname: "I_am_host",
  guestNickname: "I_am_guest",
  numberOfRounds: 4,
};
const mockGameRounds = [
  {
    id: 30,
    gameId: 14,
    hostAction: "",
    guestAction: "",
    winnerNickname: "",
    winnerAction: "",
    guestNickname: "I_am_guest",
    hostNickname: "I_am_host",
  },
  {
    id: 29,
    gameId: 14,
    hostAction: "rock",
    guestAction: "scissors",
    winnerNickname: "I_am_host",
    winnerAction: "rock",
    guestNickname: "I_am_guest",
    hostNickname: "I_am_host",
  },
  {
    id: 28,
    gameId: 14,
    hostAction: "paper",
    guestAction: "paper",
    winnerNickname: "draw",
    winnerAction: "draw",
    guestNickname: "I_am_guest",
    hostNickname: "I_am_host",
  },
];

jest.mock("../../context/Game", () => ({
  ...jest.requireActual("../../context/Game"),
  useGame: () => {
    return {
      currentGame: mockCurrentGame,
      gameRounds: mockGameRounds,
      getGameById: () => mockCurrentGame,
      fetchRoundsByGameId: () => mockGameRounds,
      createRound: () => {},
      updateRound: () => {},
    };
  },
}));
jest.mock("../../context/Player", () => ({
  ...jest.requireActual("../../context/Player"),
  usePlayer: () => {
    return {
      currentPlayer: { nickname: "dummy" },
    };
  },
}));

describe("Game container", () => {
  test("match snapshot", () => {
    const mockedProps = {
      match: { params: { id: 1 } },
    };
    const { container } = render(<Game {...mockedProps} />);
    expect(container).toMatchSnapshot();
  });

  test("renders home", () => {
    const mockedProps = {
      match: { params: { id: 14 } },
    };
    const { container } = render(<Game {...mockedProps} />);

    const gameIdLabel = screen.getByText(/Game ID/i);
    const numberOfRoundsLabel = screen.getByText(/Rounds: /i);
    const playerActionsWrp = screen.getByTestId("player-actions");
    const actionRock = screen.getByTestId("action-rock");
    const actionPaper = screen.getByTestId("action-paper");
    const actionScissors = screen.getByTestId("action-scissors");

    expect(gameIdLabel).toBeInTheDocument();
    expect(numberOfRoundsLabel).toBeInTheDocument();
    expect(playerActionsWrp).toBeTruthy();

    expect(actionRock).toBeTruthy();
    expect(actionPaper).toBeTruthy();
    expect(actionScissors).toBeTruthy();
  });
});
