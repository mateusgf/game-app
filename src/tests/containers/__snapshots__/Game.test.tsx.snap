// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game container match snapshot 1`] = `
<div>
  [{"id":30,"gameId":14,"hostAction":"","guestAction":"","winnerNickname":"","winnerAction":"","guestNickname":"I_am_guest","hostNickname":"I_am_host"},{"id":29,"gameId":14,"hostAction":"rock","guestAction":"scissors","winnerNickname":"I_am_host","winnerAction":"rock","guestNickname":"I_am_guest","hostNickname":"I_am_host"},{"id":28,"gameId":14,"hostAction":"paper","guestAction":"paper","winnerNickname":"draw","winnerAction":"draw","guestNickname":"I_am_guest","hostNickname":"I_am_host"}]
  <div
    class="player-screen"
    data-testid="player-screen"
  >
    <h1
      class="player-screen--game-id"
    >
      Game ID: 
      14
       | Rounds: 
      4
    </h1>
    <div
      class="player-actions"
      data-testid="player-actions"
    >
      <div
        class="player-actions--action-wrp"
        data-testid="action-rock"
      >
        <img
          alt="rock"
          src="rock.jpg"
        />
        <span
          class="player-actions--action-label"
        >
          rock
        </span>
      </div>
      <div
        class="player-actions--action-wrp"
        data-testid="action-paper"
      >
        <img
          alt="paper"
          src="paper.jpg"
        />
        <span
          class="player-actions--action-label"
        >
          paper
        </span>
      </div>
      <div
        class="player-actions--action-wrp"
        data-testid="action-scissors"
      >
        <img
          alt="scissors"
          src="scissors.jpg"
        />
        <span
          class="player-actions--action-label"
        >
          scissors
        </span>
      </div>
    </div>
    <hr
      class="player-screen--divider"
    />
    <h3>
      Rounds:
    </h3>
    <table
      border="1"
      width="100%"
    >
      <thead>
        <tr>
          <th>
            Round
          </th>
          <th>
            I_am_host
          </th>
          <th>
            I_am_guest
          </th>
          <th>
            Winner
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            1
          </td>
          <td>
            paper
          </td>
          <td>
            paper
          </td>
          <td>
            draw
          </td>
        </tr>
        <tr>
          <td>
            2
          </td>
          <td>
            rock
          </td>
          <td>
            scissors
          </td>
          <td>
            I_am_host
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
